tests/components/matrix/test_login.py

```diff
@@ -0,0 +1,55 @@
+"""Test MatrixBot._login."""
+
+import pytest
+
+from homeassistant.components.matrix import MatrixBot
+from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
+
+from tests.components.matrix.conftest import TEST_MXID, TEST_TOKEN
+
+
+async def test_login(matrix_bot: MatrixBot, mock_client):
+    """Test various good and bad login paths.
+
+    "Good" test password is configured by default from config.
+    """
+
+    # Test good login using good password and missing access token.
+    await matrix_bot._login()
+    assert matrix_bot._client.logged_in
+
+    # Test good login using a good password and bad stored access_token.
+    matrix_bot._access_tokens = {TEST_MXID: "WrongToken"}
+    await matrix_bot._login()
+    assert matrix_bot._client.logged_in
+
+    # Test good login using a good access_token and bad password.
```

Comment:
I suggest using parameterized tests for these different use cases as it makes the test more readable.

I need help understanding how the password is now wrong without any change. As in the lines above, it is valid.

Reply:
I agree, it was a bit opaque/I relied on some implicit operations.

I've refactored the tests to be parameterized, and the explicit marking of what password/access tokens are provided are now entirely explicit